### ISSUE TITLE
[css-fonts] Clarification font-variant-emoji should not affect characters `0-9#*`

### ISSUE THREAD

**Issue by yisibl** [w3c/csswg-drafts#11014]
> Only the code points [listed](http://www.unicode.org/emoji/charts/emoji-variants.html) by Unicode as contributing to a Unicode emoji presentation sequence are affected by this property. Within this CSS specification, these characters are referred to as Emoji Presentation Participating Code Points. This property has no effect on any other characters.

Consider the following use case:

```html
data:text/html;charset=UTF-8,<!doctype html>
<style>
  .test { font-size: 52px; }
  .test { font-variant-emoji: text; font-family: system-ui, sans-serif; }
</style> 

<div class="test">⬆379 kB/s</div>
```

![image](https://github.com/user-attachments/assets/64fd0c29-d1b9-4ea1-8d5c-d736d4adc291)

The author only wants to change the style of the arrow (⬆︎), but when font-variant-emoji: emoji is applied, the font-family of the number `379` also becomes the font-family of the Emoji. this is something the author doesn't want.

Although, [UTS#51](https://www.unicode.org/reports/tr51/#Identification) states that `0-9#*` [have the `Emoji` property,](https://www.unicode.org/Public/16.0.0/ucd/emoji/emoji-data.txt#:~:text=E0.0%20%20%20%5B1%5D%20(*%EF%B8%8F)%20%20%20%20%20%20%20asterisk-,0030..0039,-%3B%20Emoji%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20E0.0%20%20%5B10) I think they should be excluded from font-variant-emoji.

```
emoji_character := \p{Emoji}
https://www.unicode.org/reports/tr51/#def_emoji_character
```

Since Chrome has shipped, we need to clarify this in the CSS specification as soon as possible.

I've already filed a bug for Chrome: https://issues.chromium.org/issues/369781730

cc @svgeesus @drott @jfkthame @markusicu

**Comment by drott** [comment]
IMO there is no particular need or utility in the CSS spec deviating from the Unicode spec. 

> Only the code points [listed](http://www.unicode.org/emoji/charts/emoji-variants.html) by Unicode as contributing to a Unicode emoji presentation sequence are affected by this property. Within this CSS specification, these characters are referred to as Emoji Presentation Participating Code Points. This property has no effect on any other characters.

Digits, * and # belong toe presentation sequences. They have character class Emoji=Yes, EmojiPresentation=No. Excluding them from `font-variant-emoji` would break the property.

It seems to me the clarification you desire is that the styled font should not switch for value `text` of the `font-variant-emoji` when the styled `font-family` that is selected for the text can cover these characters. This is what we're discussing and planning to fix in the bug.

**Comment by yisibl** [comment]
IMO, `font-variant-emoji: text/emoji` should not change the font-family of `0-9#*`, but should always use `font-family: Arial` as set by the author.

If the author wants to use an Emoji font, it should be used with vs15 or vs16 (e.g. `U+0039, U+FE0F`). Otherwise this will break the page styling and cause the confusion in my screenshot above.

If you don't want to change the font-family of a digit, the author has to wrap the specific Emoji in a new element via an element that becomes:

```
data:text/html;charset=UTF-8,<!doctype html>
<style>
  .test { font-size: 52px; font-family: Arial; }
  .emoji { font-variant-emoji: emoji; }
</style>
<div class="test"><span class="emoji">⬆</span> 379 kB/s</div>
```

It would be horrible if, in order to change the styling, we had to wrap each emoji with a separate element. And there are times when we can't change the HTML structure of that content.

This greatly reduces the ergonomics of font-variant-emoji.

**Comment by drott** [comment]
Yes, for value `text` the family should not change, and the spec does not suggest otherwise. For value `emoji` it is the purpose of the property to upgrade to emoji presentation, and render glyphs as if they were followed by VS16. 

> If the author wants to use an Emoji font, it should be used with vs15 or vs16 (e.g. U+0039, U+FE0F). Otherwise this will break the page styling and cause the confusion in my screenshot above.

Yes, if an emoji upgrade is not intended or desired, `font-variant-emoji:` still has values `unicode` and `normal` and I think the value `unicode` matches what you want here.

**Comment by yisibl** [comment]
> For value emoji it is the purpose of the property to upgrade to emoji presentation, and render glyphs as if they were followed by VS16.

My point is that this scenario makes no sense for Digits to render as emoji fonts, and I've never seen a web page use it that way.

> font-variant-emoji: still has values unicode and normal and I think the value unicode matches what you want here.

Still the same problem I mentioned, this requires a change in the HTML structure.

Also, the `unicode` keyword changes the font-family.

```html
data:text/html;charset=UTF-8,<!doctype html>
<style>
  .test { font-size: 52px; font-family: Arial; }
  .normal { font-variant-emoji: normal; }
  .emoji { font-variant-emoji: emoji; }
  .text { font-variant-emoji: text; }
  .unicode { font-variant-emoji: unicode; }
</style>

<div class="test normal">0123456789*</div>
<div class="test unicode">0123456789*</div>
<div class="test emoji">0123456789*</div>
<div class="test text">0123456789*</div>
```

![image](https://github.com/user-attachments/assets/0496127e-187a-4dc0-aa76-e6fd04a0ae72)

**Comment by jfkthame** [comment]
> Yes, if an emoji upgrade is not intended or desired, `font-variant-emoji:` still has values `unicode` and `normal` and I think the value `unicode` matches what you want here.

`font-variant-emoji: unicode` doesn't achieve what @yisibl wants here, because the up-arrow is U+2B06. Checking Unicode's emoji-data.txt, it has the Emoji property:

> 2B05..2B07    ; Emoji                # E0.6   [3] (⬅️..⬇️)    left arrow..down arrow

but does not have Emoji_Presentation:

> 27BF          ; Emoji_Presentation   # E1.0   [1] (➿)       double curly loop
> 2B1B..2B1C    ; Emoji_Presentation   # E0.6   [2] (⬛..⬜)    black large square..white large square

so its default presentation will be the `text` form.

In terms of the Unicode Emoji and Emoji_Presentation properties, U+2B06 UPWARDS BLACK ARROW has the same properties as the ASCII digits, so it is expected that `font-variant-emoji` will affect them in the same way.

**Comment by drott** [comment]
> Also, the `unicode` keyword changes the font-family.

It shouldn't and is likely be covered by the fix for https://issues.chromium.org/issues/369781730 which should be coming to Canary soon, please check again once we have a release with the fix and let's discuss implementation issue on the Chromium tracker, not here.

**Comment by yisibl** [comment]
We should distinguish between Unicode property and what is actually needed in CSS. The Unicode may have a lot of legacy compatibility issues that make it impossible to adjust `Emoji` property, but CSS has a chance to fix them.

From the user's point of view, standalone Digits should not be in the scope of Emoji.

For example, in the popular npm package [emoji-regex](https://www.npmjs.com/package/emoji-regex), Digits will return false.

**Comment by tursunova** [comment]
> My point is that this scenario makes no sense for Digits to render as emoji fonts, and I've never seen a web page use it that way.

You are correct that in this scenario using coloured (emoji style) glyphs for digits doesn't make sense, but some emoji fonts have coloured glyphs for ASCII digits, so I assume there might be some use cases when authors might want to use emoji digits, don't you think so? If you want to make standalone digits as not emojis, then, I assume, you won't be able to use coloured emoji digits at all since you would not be able to even apply VS16 to change the presentation to emoji, isn't that right?

**Comment by yisibl** [comment]
> but some emoji fonts have coloured glyphs for ASCII digits, so I assume there might be some use cases when authors might want to use emoji digits, don't you think so? 

Usually, we have the following ways to solve it:

1. Use `@font-face`(Suppose the font `Nabla-Regular` is available locally).
But more often than not we use `webfont`, which ensures consistent styling across all OS.
There are many such icon libraries on iconfont: 
https://www.iconfont.cn/collections/detail?spm=a313x.collections_index.i1.d9df05512.4aaa3a81DONmCk&cid=7450
https://www.iconfont.cn/collections/detail?spm=a313x.collections_index.i1.d9df05512.4aaa3a81DONmCk&cid=21232

```css
@font-face {
  font-family: "color_digits";
  src: local("Nabla-Regular");
  unicode-range: U+30-39;
}
.test {
  font-family: "color_digits";
  font-size: 36px;
}
```

2. Use VS15 or VS16 (e.g. `U+0039 + U+FE0E` or `U+0039 + U+FE0F`)

Overall, I think the usage you're talking about is almost non-existent in actual websites. There seems to be no reason not to use webfont.

> you won't be able to use coloured emoji digits at all since you would not be able to even apply VS16 to change the presentation to emoji, isn't that right?

Sorry, I don't understand what that means.

**Comment by Crissov** [comment]
The special thing with digits is that VS16 alone does not make them RGI emojis, because they also need U+20E3 COMBINING ENCLOSING KEYCAP at the end of the sequence.

**Comment by yisibl** [comment]
@jfkthame

> It seems to me that if the author wants an emoji-style up-arrow here, but does not want the rendering of the following digits to be affected, the right approach would be to encode the arrow as <U+2B06, U+FE0F>. That's how you control the presentation of a specific character.

Often the text entered is not under the control of the CSS author, such as when it is entered by a third party user from a CMS.

We can't expect third-party users to learn how to enter a `Variation Selector`, or even just copy and paste an emoji character from somewhere else.

That's why CSS authors need to change it to emoji style with `font-variant-emoji: emoji`, which I believe is one of the most common scenarios where this property is used.

> Applying `font-variant-emoji` to a wider range of text inherently risks affecting other characters besides the intended one.

The purpose of this issue is to minimize such risks, and digits are the most obvious ones. CSS should not give authors unexpected negative surprises.

**Comment by yisibl** [comment]
Is this worth adding to the agenda? I'd quite like to see this progress.

**Comment by yisibl** [comment]
With the release of `font-variant-emoji` in [Chrome 131](https://chromestatus.com/feature/6566092561973248), Twitter has received a very high number of reports of users with anomalies in the display of digits.

Although Chrome quickly [released a patch](https://chromium-review.googlesource.com/c/chromium/src/+/6038776) to fix it, it clearly increased the complexity of the browser engine's implementation.

So both for users and browser engines, we should avoid this situation at the specification level.

I reported this issue more than a month ago, but unfortunately I wasn't able to change the brokenness of the Web.

- https://x.com/kousakianju/status/1860605621546177010
- https://x.com/mika2nd/status/1860566465541472314
- https://x.com/kagekoyushi/status/1860502867280286146
- https://x.com/Bach1k0_2nd/status/1860120074335715789
- https://x.com/entrance_ta2ya/status/1859966064433717515
- https://x.com/TORIOKI_SENYO/status/1859961315105862070
- https://x.com/entrance_ta2ya/status/1859004331439387097
- https://x.com/suita7116/status/1857749449466065379
- https://x.com/miyuri_ku/status/1855612942487597196

![image](https://github.com/user-attachments/assets/8c6117c6-3cc6-4b09-8139-8ff203942204)

![image](https://github.com/user-attachments/assets/015bb091-ae47-42f3-aab9-597ad4020aa6)

![image](https://github.com/user-attachments/assets/06364374-d690-4f6d-a873-a162dd1e3438)

**Comment by yisibl** [comment]
> Yes, if an emoji upgrade is not intended or desired, `font-variant-emoji:` still has values `unicode` and `normal` and I think the value `unicode` matches what you want here.

The `unicode` keyword doesn't turn all emoji into color style, that's why we need to set it to the `emoji` keyword.

In general, for the need to change digits to a colored font itself, CSS authors usually use Webfont or `VS16` instead of `font-variant-emoji` (See https://github.com/w3c/csswg-drafts/issues/11014#issuecomment-2402859557). `font-variant-emoji: emoji` should not change the font-family of a digit at any time.

**Comment by drott** [comment]
> The `unicode` keyword doesn't turn all emoji into color style, that's why we need to set it to the emoji keyword.

You can use the `emoji` keyword if you use the `font-variant-emoji` property correctly and with the right granularity. Or you can use VS15/VS16 with the `unicode` mode of the property.

From your original report, and as @jfkthame explains above, one correct way to upgrade the arrow in `⬆379 kB/s` is to place a span around the arrow and not apply `font-variant-emoji: emoji` to the full `<div>` element, another correct way is to use VS16 for the arrow and set `font-variant-emoji` to `unicode`. 

> `font-variant-emoji: emoji` should not change the font-family of a digit at any time.

Why not? Several emoji fonts contain color digits, not only for the keycap sequences. Emoji 2.0 in Unicode introduced emoji digits.

@yisibl Are you adovcating for a specific text with exceptions for codepoint ranges in how the font-variant-emoji property is applied? In that case we would end up with something roughly saying: "use unicode properties, except for... 0-9"?

The request to exclude digits from `font-variant-emoji` would be an arbitrary Unicode exception for the web. This is a distinction that from the Chrome side we oppose to making in the CSS fonts spec. What would we do we do with `*`, and `#`, what do we do with the trademark sign, the heavy plus sign, etc. - this would make for a really odd precedent.

From the Chrome side, we have strict implementation constraints to not deviate from Unicode for character properties. We include ICU, and we use [emoji-segmenter](https://github.com/google/emoji-segmenter) - which are general purpose libraries for Unicode behavior. If we would do a web-specific special casing of numbers, this causes additional complications in using these libraries - and we do not want to override their behavior. 

* Digits have the [`Emoji=Yes` property](https://unicode.org/Public/16.0.0/ucd/emoji/emoji-data.txt) and so with a VS16 they can be upgraded to emoji presentation. 
* Multiple emoji fonts support colored digits, not only colored digits with keycaps.

@yisibl If you strongly believe standalone numbers should not be emoji, I believe this would instead be a discussion for Unicode to remove the Emoji character class from the digits, or for requests to emoji font producers to remove the colored digits from them.

**Comment by yisibl** [comment]
I'm staying up late waiting for the results of this meeting, and I'm glad some progress has been made.

Thank you all for taking the time to discuss this! 🤝🤝

I'll give a summary of some of the possible doubts:

### Why not use the `unicode` keyword?

Because it doesn't allow common emoji like `⬆︎`, `❤︎` to be in color style.

### Why not nest `<span>` in emoji characters?

Again, I want to emphasize that there is a lot of content on the web that is not under the control of CSS authors, and that is completely triangulated by user input (e.g., text typed into a comment box). The cost of nesting `span`s with numbers in this content is very high.

In the meeting above, @fantasai also expressed the same idea as mine.

> moonira: dominik mentioned in his last comment we can also use span elements on those digits to achieve the desired outcome

> fantasai: yes, but the commenter is saying that digits are commonly used and rarely do they want emoji styling. Forcing the author to put spans around every digits is a lot of extra work.
 (and might not even be possible in their system)


### Looking for a Unicode solution?

> From the Chrome side, we have strict implementation constraints to not deviate from Unicode for character properties.

@drott So, am I to understand that as long as there is a relevant property in Unicode to exclude digits, Chrome is happy to implement it?

If I understand correctly, `RGI_Emoji_Qualification` in UTS 51 seems to be a solid solution, where `0-9*#` is not included.

See: 
- https://github.com/mathiasbynens/emoji-test-regex-pattern/issues/7#issuecomment-2379868224
- https://www.unicode.org/Public/emoji/16.0/emoji-test.txt

**Comment by drott** [comment]
> Why not use the unicode keyword?
> Because it doesn't allow common emoji like ⬆︎, ❤︎ to be in color style.

It does, with VS16.

> If I understand correctly, `RGI_Emoji_Qualification` in UTS 51 seems to be a solid solution, where `0-9*#` is not included.

`RGI_Emoji_Qualification` is not suitable, because
* it is not a character property, but an enumeration of strings. (we need the former in emoji segmentation)
* is too restrictive, as it enumerates the RGI Emoji set - which is different from what can occur in fonts and as you point out, [misses certain used sequences](https://github.com/mathiasbynens/emoji-test-regex-pattern/issues/7#issuecomment-2414525703).

**Comment by nikolaGrancharov** [comment]
Hey, any progress on this one?   

I am also facing the same issue, I have to use `font-variant-emoji: emoji` to make consistant, the representation for all the emoji in the application. Problem is that I can not wrap each instance of emoji in tags, since the content is done with user input (clarification: I can not do this, without significant performance hit).   

I am not that familiar with the specification, but from reading the above, and the linked items, what I am getting is that, there are some use cases, where this is expected, and others that this is not expected. I also am seeing that Firefox dose not have this issue, and `font-variant-emoji: emoji` dose not affect numbers `0-9` and the `#`.   

It looks to me that perfect solution that solves both cases is not realistic, so maybe some type of compromise is needed.  
As much as a developer I hate having workaround solutions, maybe adding additional value for the `font-variant-emoji` that excludes the numbers and the `#` will both keep the current implementation, and also satisfy the needs for people like me, and the original poster at the same time.  Something like `font-variant-emoji:emoji-only` (i am aware that the name is not great, but trying my best here)  In this way, instead of browsers deciding what the developers mean when they use `font-variant-emoji: emoji` the developers can say to the browser exactly what they expect as result.
